Added firewalled topic

MicioBaus opened the pull request:
I added a new interface, `FirewalledTopicInterface`, that allows adding logic inside user's topic service to close unwanted connections.

I extended the current TopicDispatcher to handle this type of topic service. 

Modifications are pretty strightforward, the only issue is that:
- IF a client connects to more than one topic
- AND one of them fails
- the connection is closed, disconnecting from ALL topics

I do not know if this is the correct behaviour, it is a conservative approach though.

ProPheT777 commented:
Thank your for your contribution, could you open a PR on https://github.com/GeniusesOfSymfony/WebsocketAppDemo and target your branch via composer

gos/web-socket-bundle: '1.8.6 as dev-staging-firewall'
And add a sample handshake middleware then I can run integration against it.

ProPheT777 commented on Resources/docs/TopicSetup.md:
```diff
@@ -152,6 +152,64 @@ Where
+        if ( /* ... */ )
+        {
+            // return error string to deny access
+            return 'Unauhtorized';
```
Prefer to throw a specific exception here, it should fail so make it fail in all case, add a `try {} catch ()` on top level to handle a `FirewallRejectionException` for example.

ProPheT777 commented on Resources/docs/TopicSetup.md:
```diff
@@ -152,6 +152,64 @@ Where
+        else
+        {
+            // return null to grant access
+            return null;
```
Then return nothing if it pass, in that way, if no exception is raised, it passe, we do not have to check the return value.

ProPheT777 commented on Topic/FirewalledTopicInterface.php:
```diff
@@ -0,0 +1,22 @@
+     *
+     * @return string|null        $error
+     */
+    public function checkConnection(ConnectionInterface $conn = null, Topic $topic, WampRequest $request, $payload = null, $exclude = null, $eligible = null, $provider = null);
```
Not a huge fan of method named `checkConnection`, simply call it `secure` ?

ProPheT777 commented on Topic/FirewalledTopicInterface.php:
```diff
@@ -0,0 +1,22 @@
+use Gos\Bundle\WebSocketBundle\Router\WampRequest;
+use Ratchet\Wamp\Topic;
+
+interface FirewalledTopicInterface
```
Prefer `SecuredTopicInterface` It's more explicit intent

ProPheT777 commented on Server/App/Dispatcher/TopicDispatcher.php:
```diff
@@ -130,11 +131,30 @@ public function onPublish(ConnectionInterface $conn, Topic $topic, WampRequest $
+                {
+                    $error = $appTopic->checkConnection($conn, $topic, $request, $payload , $exclude , $eligible , $provider );
+                    
+                    if ( !is_null($error) )
```
Then this check can be removed

ProPheT777 commented on Server/App/Dispatcher/TopicDispatcher.php:
```diff
@@ -130,11 +131,30 @@ public function onPublish(ConnectionInterface $conn, Topic $topic, WampRequest $
 
+                if ( $appTopic instanceof FirewalledTopicInterface )
+                {
+                    $error = $appTopic->checkConnection($conn, $topic, $request, $payload , $exclude , $eligible , $provider );
```
As said previously, we do not have to check the return result, just try catch, if catch, a firewall exception, close the connection, otherwise, it's ok

ProPheT777 commented:
Thank your for your contribution, could you open a PR on https://github.com/GeniusesOfSymfony/WebsocketAppDemo and target your branch via composer

gos/web-socket-bundle: '1.8.6 as dev-staging-firewall'
And add a sample handshake middleware then I can run integration against it.

I address some comments, mainly about naming and a minor check logic on firewall. It's a Great work :tada:

ProPheT777 commented:
Why did you close this PR ? you should add your commit where you change stuff to this branch @MicioBaus it can't work otherwise :p

I let you reopen it and repush on it

MicioBaus commented:
Answer is simple: I had no idea what I was doing.
I created a new branch, deleted another one, merged, rebased, focused on white spaces like that was an issue :)

I re-opened the firewall branch, and set it up to date with latest changes. Please help me if I need to do anything else, or I will mess up something for sure ;)

MicioBaus commented:
Any news on this? Is everything all right or is something missing on my side?

GuillaumeSTEIN commented:
Hi, this sound good to me, can someone add this ? thanks

GuillaumeSTEIN commented:
Can you let me fix the errors? I really would like to this this PR happen.

ProPheT777 commented:
thanks @MicioBaus
